sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastRangeJoin.scala

```diff
@@ -0,0 +1,411 @@
+ * TODO This join will maintain sort order. The build side rows will also be added in a lower
+ *      bound sorted fashion.
+ */
+@DeveloperApi
+case class BroadcastRangeJoin(
+    leftKeys: Seq[Expression],
+    rightKeys: Seq[Expression],
+    equality: Seq[Boolean],
+    buildSide: BuildSide,
+    left: SparkPlan,
+    right: SparkPlan)
+  extends BinaryNode {
+
+  private[this] lazy val (buildPlan, streamedPlan) = buildSide match {
+    case BuildLeft => (left, right)
+    case BuildRight => (right, left)
+  }
+
+  private[this] lazy val (buildKeys, streamedKeys) = buildSide match {
+    case BuildLeft => (leftKeys, rightKeys)
+    case BuildRight => (rightKeys, leftKeys)
+  }
+
+  override def output: Seq[Attribute] = left.output ++ right.output
+
+  @transient
+  private[this] lazy val buildSideKeyGenerator: Projection =
+    newProjection(buildKeys, buildPlan.output)
+
+  @transient
+  private[this] lazy val streamSideKeyGenerator: () => MutableProjection =
+    newMutableProjection(streamedKeys, streamedPlan.output)
+
+  private[this] val timeout: Duration = {
+    val timeoutValue = sqlContext.conf.broadcastTimeout
+    if (timeoutValue < 0) {
+      Duration.Inf
+    } else {
+      timeoutValue.seconds
+    }
+  }
+
+  // Construct the range index.
+  @transient
+  private[this] val indexBroadcastFuture = future {
+    // Deal with equality.
+    val Seq(allowLowEqual: Boolean, allowHighEqual: Boolean) = buildSide match {
+      case BuildLeft => equality.reverse
+      case BuildRight => equality
+    }
+
+    // Get the ordering for the datatype.
+    val ordering = TypeUtils.getOrdering(buildKeys.head.dataType)
+
+    // Note that we use .execute().collect() because we don't want to convert data to Scala types
+    // TODO find out if the result of a sort and a collect is still sorted.
+    val eventifier = RangeIndex.toRangeEvent(buildSideKeyGenerator, ordering)
+    val events = buildPlan.execute().map(_.copy()).collect().flatMap(eventifier)
+
+    // Create the index.
+    val index = RangeIndex.build(ordering, events, allowLowEqual, allowHighEqual)
+
+    // Broadcast the index.
+    sparkContext.broadcast(index)
+  }(BroadcastRangeJoin.broadcastRangeJoinExecutionContext)
+
+  override def doExecute(): RDD[InternalRow] = {
+    // Construct the range index.
+    val indexBC = Await.result(indexBroadcastFuture, timeout)
+
+    // Iterate over the streaming relation.
+    streamedPlan.execute().mapPartitions { stream =>
+      new Iterator[InternalRow] {
+        private[this] val index = indexBC.value
+        private[this] val streamSideKeys = streamSideKeyGenerator()
+        private[this] val join = new JoinedRow2 // TODO create our own join row...
+        private[this] var row: InternalRow = EmptyRow
+        private[this] var iterator: Iterator[InternalRow] = Iterator.empty
+
+        override final def hasNext: Boolean = {
+          var result = iterator.hasNext
```

Comment:
Multiple calls to hasNext shouldn't be a problemen. Granted the first call can have a side effect (updating the state of the iterator), but the subsequent ones won't.

A problem will occur when the next is called without calling hasNext first. I was inspired by the HashedRelation class in the same package when writing this.